src/ucp/proto/proto_debug.c

```diff
@@ -382,6 +430,207 @@ void ucp_proto_select_info_str(ucp_worker_h worker,
+    ucs_array_init_dynamic(&perf_node->children);
+    ucs_vsnprintf_safe(perf_node->desc, sizeof(perf_node->desc), desc_fmt, ap);
+
+    return perf_node;
+}
+
+static void ucp_proto_perf_node_free(ucp_proto_perf_node_t *perf_node)
+{
+    ucp_proto_perf_node_t **child_elem;
+
+    /* Delete children recursively */
+    ucs_array_for_each(child_elem, &perf_node->children) {
+        ucp_proto_perf_node_deref(child_elem);
+    }
+    ucs_array_cleanup_dynamic(&perf_node->children);
+
+    if (perf_node->type == UCP_PROTO_PERF_NODE_TYPE_DATA) {
+        ucs_array_cleanup_dynamic(&perf_node->data);
+    }
+
+    ucs_free(perf_node);
+}
+
+#define UCP_PROTO_PERF_NODE_NEW(_type, _name, _desc_fmt) \
+    ({ \
+        ucp_proto_perf_node_t *__perf_node; \
+        va_list __ap; \
+        \
+        va_start(__ap, _desc_fmt); \
+        __perf_node = ucp_proto_perf_node_new( \
+                UCP_PROTO_PERF_NODE_TYPE_##_type, _name, _desc_fmt, __ap); \
+        va_end(__ap); \
+        \
+        if (__perf_node == NULL) { \
+            return NULL; \
+        } \
+        \
+        __perf_node; \
+    })
+
+ucp_proto_perf_node_t *
+ucp_proto_perf_node_new_data(const char *name, const char *desc_fmt, ...)
+{
+    ucp_proto_perf_node_t *perf_node;
+
+    perf_node = UCP_PROTO_PERF_NODE_NEW(DATA, name, desc_fmt);
+    ucs_array_init_dynamic(&perf_node->data);
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *ucp_proto_perf_node_new_select(const char *name,
+                                                      unsigned selected_child,
+                                                      const char *desc_fmt, ...)
+{
+    ucp_proto_perf_node_t *perf_node;
+
+    perf_node = UCP_PROTO_PERF_NODE_NEW(SELECT, name, desc_fmt);
+    perf_node->selected_child = selected_child;
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *
+ucp_proto_perf_node_new_compose(const char *name, const char *desc_fmt, ...)
+{
+    return UCP_PROTO_PERF_NODE_NEW(COMPOSE, name, desc_fmt);
+}
+
+void ucp_proto_perf_node_ref(ucp_proto_perf_node_t *perf_node)
+{
+    if (perf_node != NULL) {
+        ++perf_node->refcount;
+    }
+}
+
+void ucp_proto_perf_node_deref(ucp_proto_perf_node_t **perf_node_p)
+{
+    ucp_proto_perf_node_t *perf_node = *perf_node_p;
+
+    if (perf_node == NULL) {
+        return;
+    }
+
+    ucs_assertv(perf_node->refcount > 0, "perf_node=%p name='%s' desc='%s'",
+                perf_node, perf_node->name, perf_node->desc);
+    --perf_node->refcount;
+    if (perf_node->refcount == 0) {
+        ucp_proto_perf_node_free(perf_node);
+        *perf_node_p = NULL;
+    }
+}
+
+static void
+ucp_proto_perf_node_append_child(ucp_proto_perf_node_t *perf_node,
+                                 ucp_proto_perf_node_t *child_perf_node)
+{
+    ucs_status_t status;
+
+    status = ucs_array_append(ucp_proto_perf_node, &perf_node->children);
+    if (status != UCS_OK) {
+        ucs_diag("failed to add perf node child");
+        return;
+    }
```

Comment:
maybe replace by `ucs_assertv_always(status != UCS_OK, "failed to add perf node child: %s", ucs_status_string(status))`
normally it shouldn't happen

Reply:
in perf_node we ignore errors related to memory allocation since the applications can work without having a performance information tree.